frame/staking/src/lib.rs

```diff
@@ -1517,106 +1471,90 @@ impl<T: Trait> Module<T> {
+		// nominators will get.
+		let validator_total_reward_part = Perbill::from_rational_approximation(
+			validator_reward_points,
+			total_reward_points,
+		);
 
-				let nominator_exposure_part = Perbill::from_rational_approximation(
-					nominator_exposure.value,
-					validator_exposure.total,
-				);
-				let validator_point = era_reward_points.individual.get(&validator)
-					.map(|points| *points)
-					.unwrap_or_else(|| Zero::zero());
-				let validator_point_part = Perbill::from_rational_approximation(
-					validator_point,
-					era_reward_points.total,
-				);
-				reward = reward.saturating_add(
-					validator_point_part
-						.saturating_mul(Perbill::one().saturating_sub(commission))
-						.saturating_mul(nominator_exposure_part)
-				);
-			}
-		}
+		// This is how much validator + nominators are entitled to.
+		let validator_total_payout = validator_total_reward_part * era_payout;
+		// Validator first gets a cut off the top.
+		let validator_prefs = Self::eras_validator_prefs(&era, &ledger.stash);
+		let validator_commission = validator_prefs.commission;
+		let validator_commission_payout = validator_commission * validator_total_payout;
+
+		let validator_leftover_payout = validator_total_payout - validator_commission_payout;
+		// Now let's calculate how this is split to the validator.
+		let exposure = <ErasStakers<T>>::get(&era, &ledger.stash);
+		let validator_exposure_part = Perbill::from_rational_approximation(
+			exposure.own,
+			exposure.total,
+		);
+		let validator_staking_payout = validator_exposure_part * validator_leftover_payout;
 
-		if let Some(imbalance) = Self::make_payout(&nominator_ledger.stash, reward * era_payout) {
-			Self::deposit_event(RawEvent::Reward(who, imbalance.peek()));
+		// We can now make total validator payout:
+		if let Some(imbalance) = Self::make_payout(
+			&ledger.stash,
+			validator_staking_payout + validator_commission_payout
+		) {
+			Self::deposit_event(RawEvent::Reward(validator, imbalance.peek()));
 		}
 
-		Ok(())
-	}
-
-	fn do_payout_validator(who: T::AccountId, era: EraIndex) -> DispatchResult {
-		// Note: if era has no reward to be claimed, era may be future. better not to update
-		// `ledger.last_reward` in this case.
-		let era_payout = <ErasValidatorReward<T>>::get(&era)
-			.ok_or_else(|| Error::<T>::InvalidEraToReward)?;
-
-		let mut ledger = <Ledger<T>>::get(&who).ok_or_else(|| Error::<T>::NotController)?;
-		if ledger.last_reward.map(|last_reward| last_reward >= era).unwrap_or(false) {
-			return Err(Error::<T>::InvalidEraToReward.into());
+		// Lets now calculate how this is split to the nominators.
+		// Sort nominators by highest to lowest exposure, but only keep `max_payout` of them.
+		let max_payout = validator_prefs.max_payout as usize;
+		let mut nominators_exposure = exposure.others;
+		if nominators_exposure.len() > max_payout {
+			nominators_exposure.sort_unstable_by(|a, b| b.value.cmp(&a.value));
+			nominators_exposure.truncate(max_payout);
```

Comment:
I think your intent is to limit the number of stakers that the transactor pays out, but this logic here prevents people outside of `validator_prefs.max_payout` from ever claiming their reward, even if they are willing to pay a tx fee to get it. This breaks down to some weird incentives, especially for bigger validators, like setting `max_payout` to 1 so that only you nominate yourself and if someone else out-nominates you then your incentive is to misbehave and slash them.

Reply:
It would be a slightly difficult to allow for an additional payout nominator. It could not be used by the `max_payouts` people since we do not separately track which payouts they have claimed, and would be at risk of a double payout.

I could make it so it works for anyone not in the `max_payouts` group, but then the API is a little shitty since it cannot be used by everyone.

I personally don't think things like `max_payout = 1`(which is inherently a kind of flawed setting) need to be handled well, since we should discourage a validator from selecting this, but there are def solutions if you think it is necessary.

Reply:
It's really hard (impossible) to set a proper weight for this function because you don't know `max_payout` in advance for the validator in question. A validator could set this to be very high and make the nomination accounts to attack the chain.

I would rather see a solution that doesn't give the validator this power, but rather does pay out all nominators by:
1. Ordering nominators by exposure.
2. Subtracting some fee from the reward.
3. Truncate the list by only paying out those for whom reward is greater than zero (or some threshold like `ED/10`)

Reply:
Why `max_payout` - why not just payout all? (Excuse my ignorance, sometimes I need a "lemme ask really stupid question" forum)

Reply:
I think we are always asking you stupid questions. We are trying to avoid the situation with too many state updates in a single transaction. But I would rather pay out all as long as there is a way to decide if an account is worth paying out.

Reply:
> It's really hard (impossible) to set a proper weight for this function because you don't know max_payout in advance for the validator in question. A validator could set this to be very high and make the nomination accounts to attack the chain.

This is not really correct. We will likely refactor all extrinsics to include information like this. So in the next iteration, `max_payouts` will be a value that has to be input into the extrinsic, and matched against what is in storage. If it does not match correctly, then we do nothing.

We will charge the appropriate fee for that action.

Reply:
> Why max_payout - why not just payout all? (Excuse my ignorance, sometimes I need a "lemme ask really stupid question" forum)

This is an attack vector. A malicious actor can make many many micro nominators to prevent a validator from doing a payout with a reasonable fee.

Reply:
> This is not really correct. We will likely refactor all extrinsics to include information like this. So in the next iteration, `max_payouts` will be a value that has to be input into the extrinsic, and matched against what is in storage. If it does not match correctly, then we do nothing.
> 
> We will charge the appropriate fee for that action.

I still don't see how this is a superior solution than deducting a fee from each staker's rewards and truncating the list based on that. You are giving validators the power to truncate the list of rewarded accounts rather than using an economic mechanism. Charging a fee per rewarded account would basically add a cost to nominators for dividing their account up into smaller accounts that all nominate the same target, which is exactly the behavior we want to disincent. Just giving this param to validators seems like a hack to me.

Reply:
Ah, I see what you mean, and generally it makes sense, but our weight/fee system isnt quite smart enough to handle this.

The underlying weight of the call needs to appropriately measured/reported for the purposes of forming blocks, and our fee mechanism depends directly on this weight. So even if we do try to "take fees from the rewards" we will still need to increase the weight of the call proportional to the number of people paid out, and ultimately this would need to be known ahead of time.